File: kyc-deal/src/engine/brain.ts
/**
 * The KYC Brain. Every finalized answer is folded into a library keyed by the
 * normalized question. As the corpus grows, recurring questions accumulate
 * the same value again and again — optionality (the count of distinct
 * answers) stays low and convergence (the dominant answer's share) climbs. So
 * the more questionnaires a client has answered, the more the next one
 * answers itself.
 */

import { getDb, genId } from '../db/db.js';
import type { BrainOption } from '../types.js';

const STOP = new Set([
  'the', 'a', 'an', 'of', 'for', 'to', 'is', 'are', 'please', 'kindly', 'provide', 'state',
  'your', 'you', 'and', 'or', 'in', 'on', 'with', 'this', 'that', 'any', 'each', 'all', 'we',
  'do', 'does', 'has', 'have', 'what', 'which', 'name', 'list',
]);

/** Collapse trivial differences in phrasing so "Please state the LEI of the
 *  entity" and "Provide the entity's LEI" land on the same key. */
export function normalizeQuestion(prompt: string): string {
  const tokens = prompt
    .toLowerCase()
    .replace(/[^a-z0-9%\s]/g, ' ')
    .split(/\s+/)
    .filter((t) => t && !STOP.has(t));
  return tokens.sort().join(' ').trim();
}

interface Variant {
  value: string;
  count: number;
}

function parseVariants(json: string): Variant[] {
  try {
    const v = JSON.parse(json) as Variant[];
    return Array.isArray(v) ? v : [];
  } catch {
    return [];
  }
}

function variantsToOptions(variants: Variant[]): BrainOption[] {
  const total = variants.reduce((s, v) => s + v.count, 0) || 1;
  return [...variants]
    .sort((a, b) => b.count - a.count)
    .map((v) => ({ value: v.value, timesUsed: v.count, share: v.count / total }));
}

/** Prior answers to a question, most-used first. */
export function getBrainOptions(prompt: string): BrainOption[] {
  const db = getDb();
  const norm = normalizeQuestion(prompt);
  const row = db.prepare(`SELECT variants_json FROM answer_library WHERE question_norm = ?`).get(norm) as { variants_json: string } | undefined;
  if (!row) return [];
  return variantsToOptions(parseVariants(row.variants_json));
}

/** Fold a finalized answer into the library. */
export function recordFinalizedAnswer(prompt: string, kind: string, value: string): void {
  const v = value.trim();
  if (!v) return;
  const db = getDb();
  const norm = normalizeQuestion(prompt);
  const row = db.prepare(`SELECT id, variants_json, times_used FROM answer_library WHERE question_norm = ?`).get(norm) as
    | { id: string; variants_json: string; times_used: number }
    | undefined;

  if (!row) {
    const variants: Variant[] = [{ value: v, count: 1 }];
    db.prepare(
      `INSERT INTO answer_library (id, question_norm, sample_prompt, question_kind, variants_json, times_used, last_used)
       VALUES (?, ?, ?, ?, ?, 1, datetime('now'))`,
    ).run(genId('lib'), norm, prompt.trim(), kind, JSON.stringify(variants));
    return;
  }

  const variants = parseVariants(row.variants_json);
  const key = (s: string) => s.toLowerCase().replace(/\s+/g, ' ').trim();
  const existing = variants.find((x) => key(x.value) === key(v));
  if (existing) existing.count++;
  else variants.push({ value: v, count: 1 });
  db.prepare(`UPDATE answer_library SET variants_json = ?, times_used = times_used + 1, last_used = datetime('now') WHERE id = ?`).run(
    JSON.stringify(variants),
    row.id,
  );
}

export interface BrainEntry {
  id: string;
  prompt: string;
  kind: string;
  timesUsed: number;
  optionality: number; // distinct answers seen
  dominantShare: number; // share of the most common answer
  options: BrainOption[];
  lastUsed: string;
}

export function listBrain(): BrainEntry[] {
  const db = getDb();
  const rows = db
    .prepare(`SELECT id, sample_prompt, question_kind, variants_json, times_used, last_used FROM answer_library ORDER BY times_used DESC, last_used DESC`)
    .all() as { id: string; sample_prompt: string; question_kind: string; variants_json: string; times_used: number; last_used: string }[];
  return rows.map((r) => {
    const options = variantsToOptions(parseVariants(r.variants_json));
    return {
      id: r.id,
      prompt: r.sample_prompt,
      kind: r.question_kind,
      timesUsed: r.times_used,
      optionality: options.length,
      dominantShare: options[0]?.share ?? 0,
      options,
      lastUsed: r.last_used,
    };
  });
}

export interface BrainStats {
  questions: number;
  finalizedAnswers: number;
  settled: number; // questions with a single dominant answer (optionality 1)
  avgOptionality: number;
}

export function brainStats(): BrainStats {
  const entries = listBrain();
  const questions = entries.length;
  const finalizedAnswers = entries.reduce((s, e) => s + e.timesUsed, 0);
  const settled = entries.filter((e) => e.optionality === 1 && e.timesUsed > 1).length;
  const avgOptionality = questions ? entries.reduce((s, e) => s + e.optionality, 0) / questions : 0;
  return { questions, finalizedAnswers, settled, avgOptionality };
}

File: dealproof-py/app/engine/brain.py
"""KYC Brain — folds finalized answers into a library keyed by the normalized
question; recurring questions converge on a dominant answer."""
import json
import re
from ..db import db, gen_id, one

STOP = {"the", "a", "an", "of", "for", "to", "is", "are", "please", "kindly", "provide", "state",
        "your", "you", "and", "or", "in", "on", "with", "this", "that", "any", "each", "all", "we",
        "do", "does", "has", "have", "what", "which", "name", "list"}


def normalize_question(prompt: str) -> str:
    toks = [t for t in re.sub(r"[^a-z0-9%\s]", " ", prompt.lower()).split() if t and t not in STOP]
    return " ".join(sorted(toks)).strip()


def get_brain_options(prompt: str):
    key = normalize_question(prompt)
    with db() as con:
        row = one(con, "SELECT * FROM answer_library WHERE norm_question=?", (key,))
    if not row:
        return []
    variants = json.loads(row["variants_json"] or "[]")
    total = row["total"] or sum(v["count"] for v in variants) or 1
    variants.sort(key=lambda v: -v["count"])
    return [{"value": v["value"], "timesUsed": v["count"], "share": v["count"] / total} for v in variants]


def record_finalized_answer(prompt: str, value: str) -> None:
    if not value.strip():
        return
    key = normalize_question(prompt)
    with db() as con:
        row = one(con, "SELECT * FROM answer_library WHERE norm_question=?", (key,))
        if row:
            variants = json.loads(row["variants_json"] or "[]")
            for v in variants:
                if v["value"] == value:
                    v["count"] += 1
                    break
            else:
                variants.append({"value": value, "count": 1})
            con.execute("UPDATE answer_library SET variants_json=?, total=total+1, sample_prompt=?, updated_at=datetime('now') WHERE id=?",
                        (json.dumps(variants), prompt, row["id"]))
        else:
            con.execute("INSERT INTO answer_library (id, norm_question, sample_prompt, variants_json, total) VALUES (?,?,?,?,1)",
                        (gen_id("lib"), key, prompt, json.dumps([{"value": value, "count": 1}])))


# answer-cell values that aren't real answers (don't learn these)
NON_ANSWERS = {"", "n/a", "na", "n.a.", "none", "please select", "-", "–", "tbd", "yes/no", "y/n"}


def learn_from_document(filename: str, content: bytes, raw_text: str = "") -> dict:
    """Seed the Brain from a PAST, already-answered questionnaire: extract each
    question + its given answer and fold it in. Existing cases entered this way
    become the basis the connectors (Quantium / YSolutions) only verify.
    A Transparenzregister extract is recognised and handled as what it is: the
    beneficial owners are learned under the company they belong to.
    Returns {total, learned, skipped[, kind, company]}."""
    from .intake import parse_document, docx_qa_pairs
    from . import transparency, chartpdf
    # a drawn structure chart is structural data, not Q&A — recognise it and
    # point the user to the right place instead of "learning 0 answers"
    if (filename or "").lower().endswith(".pdf") and content and chartpdf.is_chart_pdf(content):
        spec = chartpdf.extract_spec(content) or {"entities": [], "edges": []}
        return {"total": len(spec["edges"]), "learned": len(spec["entities"]),
                "skipped": 0, "kind": "structurechart", "company": ""}
    if raw_text and transparency.is_tr_extract(raw_text):
        ubos = transparency.extract_ubos(raw_text)
        if ubos:                                   # a blank UBO form falls through
            company = transparency.tr_company(raw_text)
            transparency.learn_ubos_into_brain(ubos, company=company)
            return {"total": len(ubos), "learned": len(ubos),
                    "skipped": 0, "kind": "transparenzregister", "company": company}
    name = (filename or "").lower()
    if name.endswith(".docx") and content:
        pairs = docx_qa_pairs(content)
    else:  # PDF (table/markdown captures the answer cell) and anything else
        pairs = [(q["prompt"], q.get("answer", "")) for q in parse_document(raw_text, filename, content)]
    learned = skipped = 0
    for prompt, ans in pairs:
        ans = (ans or "").strip()
        if prompt and ans and ans.lower() not in NON_ANSWERS:
            record_finalized_answer(prompt, ans)
            learned += 1
        else:
            skipped += 1
    return {"total": len(pairs), "learned": learned, "skipped": skipped}


def brain_stats():
    with db() as con:
        lib = [dict(r) for r in con.execute("SELECT * FROM answer_library ORDER BY total DESC").fetchall()]
    entries = []
    for r in lib:
        variants = sorted(json.loads(r["variants_json"] or "[]"), key=lambda v: -v["count"])
        total = r["total"] or 1
        top = variants[0] if variants else {"value": "", "count": 0}
        entries.append({"id": r["id"], "prompt": r["sample_prompt"], "total": r["total"],
                        "optionality": len(variants), "convergence": top["count"] / total,
                        "answer": top["value"]})
    return entries


def update_entry(entry_id: str, value: str) -> None:
    """Correct a learned answer: the given value becomes THE canonical answer
    (all previous variants are replaced, convergence resets to 100%)."""
    value = (value or "").strip()
    if not value:
        return
    with db() as con:
        row = one(con, "SELECT total FROM answer_library WHERE id=?", (entry_id,))
        if row:
            total = max(row["total"] or 1, 1)
            con.execute("UPDATE answer_library SET variants_json=?, updated_at=datetime('now') WHERE id=?",
                        (json.dumps([{"value": value, "count": total}]), entry_id))


def delete_entry(entry_id: str) -> None:
    """Forget a learned question/answer entirely."""
    with db() as con:
        con.execute("DELETE FROM answer_library WHERE id=?", (entry_id,))
